pkg/controllers/provisioningv2/rke2/common.go

```diff
@@ -196,22 +199,37 @@ func IsOwnedByMachine(bootstrapCache rkecontroller.RKEBootstrapCache, machineNam
 }
 
 // GetServiceAccountSecretNames will return the plan secret name and service account token names
-func GetServiceAccountSecretNames(bootstrapCache rkecontroller.RKEBootstrapCache, machineName string, planSA *corev1.ServiceAccount) (string, string, error) {
+func GetServiceAccountSecretNames(bootstrapCache rkecontroller.RKEBootstrapCache, secretClient corecontrollers.SecretController, machineName string, planSA *corev1.ServiceAccount) (string, string, error) {
 	if planSA.Labels[MachineNameLabel] != machineName ||
 		planSA.Labels[RoleLabel] != RolePlan ||
 		planSA.Labels[PlanSecret] == "" {
 		return "", "", nil
 	}
 
-	if len(planSA.Secrets) == 0 {
-		return "", "", nil
+	sName := planSA.Name + "-token"
+	secret, err := secretClient.Cache().Get(planSA.Namespace, sName)
+	if err != nil {
+		if !apierror.IsNotFound(err) {
+			return "", "", err
+		}
+		sc := serviceaccounttoken.SecretTemplate(planSA)
+		secret, err = secretClient.Create(sc)
+		if err != nil {
+			if !apierror.IsAlreadyExists(err) {
+				return "", "", err
+			}
+			secret, err = secretClient.Cache().Get(planSA.Namespace, sName)
```

Comment:
Not sure I understand this logic. At this point, we already know there was an err trying to pull the secret from the cache. Did you mean this?

`secret, err = h.secretClient.Get(sa.Namespace, sName, metav1.GetOptions{})`

Reply:
We know there's an err from cache, so we create secret after that first and then retry getting it.